server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -254,6 +239,55 @@ public void putSynonymsSet(String resourceName, SynonymRule[] synonymsSet, Actio
         }));
     }
 
+    public void putSynonymRule(String synonymSetId, SynonymRule synonymRule, ActionListener<UpdateSynonymsResult> listener) {
+        checkSynonymSetExists(synonymSetId, listener.delegateFailure((l1, obj) -> {
+            try {
+                IndexRequest indexRequest = createSynonymRuleIndexRequest(synonymSetId, synonymRule).setRefreshPolicy(
+                    WriteRequest.RefreshPolicy.IMMEDIATE
+                );
+                client.index(indexRequest, l1.delegateFailure((l2, indexResponse) -> {
+                    UpdateSynonymsResult result = indexResponse.status() == RestStatus.CREATED
+                        ? UpdateSynonymsResult.CREATED
+                        : UpdateSynonymsResult.UPDATED;
+                    l2.onResponse(result);
+                }));
+            } catch (IOException e) {
+                l1.onFailure(e);
+            }
+        }));
+    }
+
+    private static IndexRequest createSynonymRuleIndexRequest(String synonymsSetId, SynonymRule synonymRule) throws IOException {
+        try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+            builder.startObject();
+            {
+                builder.field(SYNONYMS_SET_FIELD, synonymsSetId);
+                builder.field(SYNONYM_RULE_ID_FIELD, synonymRule.id());
+                builder.field(SYNONYMS_FIELD, synonymRule.synonyms());
+            }
+            builder.endObject();
+
+            return new IndexRequest(SYNONYMS_ALIAS_NAME).id(internalSynonymRuleId(synonymsSetId, synonymRule))
+                .opType(DocWriteRequest.OpType.INDEX)
+                .source(builder)
+                .id(internalSynonymRuleId(synonymsSetId, synonymRule));
+        }
+    }
+
+    private <T> void checkSynonymSetExists(String synonymSetId, ActionListener<T> listener) {
```

Comment:
I refactored this out as it will be reused by `DELETE` and `GET` actions as well; we need to check for the synonym set existence.